internal/server/data/deviceflowauthrequest.go

```diff
@@ -0,0 +1,97 @@
+package data
+
+import (
+	"time"
+
+	"gorm.io/gorm"
+
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/internal/validate"
+	"github.com/infrahq/infra/uid"
+)
+
+type deviceFlowAuthRequestTable models.DeviceFlowAuthRequest
+
+func (d deviceFlowAuthRequestTable) Table() string {
+	return "device_flow_auth_requests"
+}
+
+func (d deviceFlowAuthRequestTable) Columns() []string {
+	return []string{"id", "user_code", "device_code", "approved", "access_key_id", "expires_at", "created_at",
+		"updated_at"}
+}
+
+func (d deviceFlowAuthRequestTable) Values() []any {
+	return []any{d.ID, d.UserCode, d.DeviceCode, d.Approved, d.AccessKeyID, d.ExpiresAt, d.CreatedAt, d.UpdatedAt}
+}
+
+func (d *deviceFlowAuthRequestTable) ScanFields() []any {
+	return []any{d.ID, d.UserCode, d.DeviceCode, d.Approved, d.AccessKeyID, d.ExpiresAt, d.CreatedAt, d.UpdatedAt}
+}
+
+func validateDeviceFlowAuthRequest(dfar *models.DeviceFlowAuthRequest) error {
+	err := validate.Error{}
+	validationRules := []validate.ValidationRule{
+		validate.String("user_code", dfar.UserCode, 8, 8, validate.DeviceFlowUserCode),
+		validate.String("device_code", dfar.DeviceCode, 38, 38, validate.AlphaNumeric),
```

Comment:
This validation should be on the `api` type so that we validate the input earlier in the request flow. 

Some of these are checking values that are not user input, so a 400 error would not be correct in most of these cases, but the error being returned here will become a 400 response code.

Reply:
I'll wrap the error so it's not a 400. These values are generated server-side, so there's less value in validating them from the client, but I can include that too.